src/common/time/instant.ts

```diff
@@ -0,0 +1,143 @@
+/**
+ * @license
+ * Copyright 2019 Google LLC.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {TimeUnit} from './timeunit.js';
+
+/**
+ * An Instant represents a specific instance of universal time with a
+ * defined precision.
+ *
+ * TODO implement comparable?
+ * TODO implement other utility methods
+ */
+export class Instant {
+  private readonly secs: number;
+  private readonly nanos: number;
+  private readonly resolutionValue: TimeUnit;
+  
+  private constructor(secs: number, nanos = 0, resolution: TimeUnit) {
+    if (nanos > 1000000 || nanos < 0 || secs < 0) {
+      throw new Error('input values out of range secs=' + secs + ' nanos=' + nanos);
+    }
+    this.secs = secs;
+    this.nanos = nanos;
+    this.resolutionValue = resolution;
+  }
+
+  // TODO epochNanoseconds and epochMicroseconds require bigint
+
+  /**
+   * Returns the rounded number of milliseconds from the Epoch for this Instant..
+   */
+  get epochMilliseconds(): number {
+    return this.secs * 1000 + Math.round(this.nanos / 1000000);
+  }
+
+  /**
+   * Returns the rounded number of seconds from the Epoch for this Instant.
+   */
+  get epochSeconds(): number {
+    return this.nanos > 500000 ? this.secs + 1 : this.secs;
+  }
+
+  /**
+   * Provides the resolution of this Instant based on the inputs
+   */
+  get resolution(): TimeUnit {
+    return this.resolutionValue;
+  }
+  
+  /**
+   * Creates a new Instant from the given milliseconds.  The returned
+   * value uses TimeUnit.MILLIS as the resolution.
+   */
+  static fromEpochMilliseconds(epochMillis: number): Instant {
+    // TODO check for negative numbers.
+    const secs = Math.floor(epochMillis/1000);
+    const nanos = (epochMillis % 1000) * 1000;
+    return new Instant(secs, nanos, TimeUnit.MILLIS);
+  }
+
+  
+  /**
+   * Creates a new Instant from the given seconds.  The returned
+   * value uses TimeUnit.SECONDS as the resolution.
+   */
+  static fromEpochSeconds(epochSeconds: number): Instant {
+    return new Instant(epochSeconds, 0, TimeUnit.SECONDS);
+  }
+
+  // Provides constants for parsing differnet input lengths
+  // TODO add millisecond and finer resolution
```

Comment:
Need nano seconds as well

Reply:
"and finer" :)  -- but yeah, explicitly listing todo items instead for more accuracy.